cpp/src/arrow/acero/partition_util.h

```diff
@@ -62,7 +62,7 @@ class PartitionSort {
   template <class INPUT_PRTN_ID_FN, class OUTPUT_POS_FN>
   static void Eval(int64_t num_rows, int num_prtns, uint16_t* prtn_ranges,
                    INPUT_PRTN_ID_FN prtn_id_impl, OUTPUT_POS_FN output_pos_impl) {
-    ARROW_DCHECK(num_rows > 0 && num_rows <= (1 << 15));
+    ARROW_DCHECK(num_rows > 0 && num_rows <= ((1 << 16) - 1));
     ARROW_DCHECK(num_prtns >= 1 && num_prtns <= (1 << 15));
```

Comment:
Same here (why `1<<15`??).

Reply:
```
SwissTableForJoinBuild::Init()

  constexpr int64_t min_num_rows_per_prtn = 1 << 18;
  log_num_prtns_ =
      std::min(bit_util::Log2(dop_),
               bit_util::Log2(bit_util::CeilDiv(num_rows, min_num_rows_per_prtn)));
  num_prtns_ = 1 << log_num_prtns_;
```


According to the initialization calculation here, this num_prtns should be more than 1 << 15, so do we also adjust according to the above num_rows?

Reply:
I don't understand why that would always be the case? This is using `std::min`, not `std::max`?

Reply:
I don't know either 😭

Reply:
I believe the goal here is to use up to `dop_` partitions but only use that many if we have enough rows to justify it.  We only want to create more partitions if these partitions have `min_num_rows_per_prtn` rows.  If there are not very many rows then we use fewer partitions.

Also, we require `num_prtns_` to be a power of 2 because we are going to calculate the partition id using masking and so we need `log_num_prtns_` to tell us how many bits to use for masking.

Reply:
There seems to be no upper limit for num_prtns, so here I want to adjust it to the upper limit of num_rows, maintaining the same principle as the original two.